File: js/sunblind/layer.js
function Layer(options) {
	var self = this;
	this.element = null;
	this.lamellas = [];
	this.materials = [];
	this.context = this.sunblind.element;

	this.init();
}

Layer.prototype.init = function() {
	
};

Layer.prototype.paint = function(options) {
	for(var i = 0; i < this.lamellas.length;i++)
		this.lamellas[i].paint();
	this.context.appendChild(this.element);
};

Layer.prototype.destroy = function() {
	this.element.outerHTML = "";
};

Layer.prototype.calculate = function(options) {
	var result = 0;
	return result;
};

Layer.prototype.addMaterial = function(material) {
	for(var i = 0; i < this.materials.length; i++)
		if(this.materials[i].id == material.id) {
			this.materials[i].count++; 
			return;
		}
	material.count = 1;
	this.materials.push(material);
}

Layer.prototype.addLamella = function(lamella) {
	lamella.layer = this;
	this.lamellas.push(lamella);
};

Layer.prototype.fromJSON = function(obj) {
	var lamellas = obj.lamellas;
	for(var i = 0; i < lamellas.length; i++) {
		var lamella = new Lamella();
		lamella.fromJSON(lamellas[i]);
		this.addLamella(lamella);
	}
};

Layer.prototype.toJSON = function() {
	var lamellas = [];
	for(var i = 0; i < this.lamellas.length; i++)
		lamellas.push(this.lamellas[i].toJSON());
	return {
		"lamellas": lamellas,
		minLayerSquare: define.sunblind.MIN_LAYER_SQUARE,
		width: this.sunblind.width,
		height: this.sunblind.height
	};
};

Layer.prototype.generate = function() {

};

File: cpp/classes/sunblind/layer.cc
#ifndef LAYER_CC
#define LAYER_CC

#include <vector>
#include <map>
#include "lamella.cc"

class Layer
{
public:
    Layer() : width(0), height(0), minSquare(0) {}

    Layer(int ww, int hh, float ms)
    {
        width = ww;
        height = hh;
        minSquare = ms;
    }

    ~Layer()
    {
        for(int i = 0; i < lamellas.size(); i++)
            delete lamellas[i];
    }

    virtual float calculate() = 0;

    void addLamella(Lamella& lamella)
    {
        lamellas.push_back(&lamella);
        colors[lamella.price]++;
    }

    std::vector<Lamella*>& getLamellas()
    {
        return lamellas;
    }

protected:
    int width;
    int height;
    float minSquare;
    std::map<float, int> colors;

private:
    std::vector<Lamella*> lamellas;
};

#endif
